backtester/data/data.go

```diff
@@ -51,84 +57,132 @@ func (h *HandlerPerCurrency) GetDataForCurrency(ev common.EventHandler) (Handler
 
 // Reset returns the struct to defaults
 func (h *HandlerPerCurrency) Reset() {
+	if h == nil {
+		return
+	}
 	h.data = nil
 }
 
 // Reset loaded data to blank state
 func (b *Base) Reset() {
+	if b == nil {
+		return
+	}
 	b.latest = nil
 	b.offset = 0
 	b.stream = nil
 }
 
 // GetStream will return entire data list
-func (b *Base) GetStream() []common.DataEventHandler {
+func (b *Base) GetStream() []Event {
 	return b.stream
 }
 
 // Offset returns the current iteration of candle data the backtester is assessing
-func (b *Base) Offset() int {
+func (b *Base) Offset() int64 {
 	return b.offset
 }
 
 // SetStream sets the data stream for candle analysis
-func (b *Base) SetStream(s []common.DataEventHandler) {
+func (b *Base) SetStream(s []Event) {
 	b.stream = s
+	// due to the Next() function, we cannot take
+	// stream offsets as is, and we re-set them
+	for i := range b.stream {
+		b.stream[i].SetOffset(int64(i + 1))
+	}
 }
 
 // AppendStream appends new datas onto the stream, however, will not
 // add duplicates. Used for live analysis
-func (b *Base) AppendStream(s ...common.DataEventHandler) {
-	for i := range s {
-		if s[i] == nil {
+func (b *Base) AppendStream(s ...Event) {
+candles:
+	for x := range s {
+		if s[x] == nil || !b.equalSource(s[x]) {
```

Comment:
I'll be looking into the errors. but what do you mean "it seems like we are appending nil events"?

Reply:
If we are getting a nil Event in variadic param `s` we are either appending and unfurling an event slice which contains a nil or parsing in a nil argument.

Reply:
I still don't understand sorry. In the test `TestAppendResults`, we have an instance where we pass in a `nil` event for the purpose of being skipped.
If I also add in the scenario `hello.AppendStream()` there are no events to be appended.
Are you experiencing `nil` events being added?

Are you saying, that if I experience receiving a `nil` event amongst other valid events, I should not be applying any events at all to the stream?

Reply:
Yeah return error is what I am getting at. 
`hello.AppendStream()` Should not be called
`hello.AppendStream(nil)` Should return an error
`hello.AppendStream([]Event{Event,Event,nil}... )`  <-- should error because it should not be in that slice in the first place